### PR TITLE
feat(moonshot): implement provider with model picker, API fetching, streaming, and model metadata

### PR DESCRIPTION
**Pull request by grizmin**
## Summary

Implement the Moonshot provider from scratch, enabling full model selection, API-based model fetching, and working streaming responses.

## What was missing

The previous Moonshot implementation was non-functional:
- **No model picker** — users could not select which Moonshot model to use
- **No auto-updated model list** — models were never fetched from the Moonshot API
- **Custom model ID did not work** — entering a custom model ID produced empty streams with no output
- **Wrong handler base class** — `MoonshotHandler` extended `OpenAICompatibleHandler` (Vercel AI SDK) which has compatibility issues with the Moonshot API

### Handler refactor (`src/api/providers/moonshot.ts`)
- Switched `MoonshotHandler` from `OpenAICompatibleHandler` (Vercel AI SDK) to `OpenAiHandler` (OpenAI Node SDK)
- This is the same handler used by the generic "OpenAI Compatible" provider and by DeepSeek/Mimo
- Moonshot-specific overrides kept: `getModel()`, `processUsageMetrics()`, `addMaxTokensIfNeeded()`

### Model fetching (`src/api/providers/fetchers/moonshot.ts`)
- New `getMoonshotModels()` fetcher that calls `GET /v1/models` 
- Enriches API-returned model IDs with metadata from static model map (pricing, context window, capabilities)
- Provides fallback defaults for unknown models

### Model picker UI (`webview-ui/src/components/settings/providers/Moonshot.tsx`)
- New settings component with API key input, base URL selector, and model dropdown
- "Refresh Models" button to re-fetch the model list from the API
- Shows only API-fetched models (not merged with static list)

### Model metadata (`packages/types/src/providers/moonshot.ts`)
- Added `kimi-k2.7-code`, `kimi-k2.7-code-highspeed`, `kimi-k2.6` with correct pricing from official docs
- Fixed `kimi-k2.5` to correctly mark `supportsImages: true`

### Infrastructure
- Added `moonshot` to `dynamicProviders` list (enables model fetching pipeline)
- Added Moonshot to model caching (`modelCache.ts`)
- Added webview message handler for `requestRouterModels`

## Testing
- All 6660+ backend tests pass
- All 1432+ webview-ui tests pass
- Lint and type-check pass
- Manually verified with F5 Extension Development Host using kimi-k2.7-code

## Summary by CodeRabbit

* **New Features**
  * Added Moonshot as a supported dynamic provider, including router model refresh and improved model selection using router-provided model definitions.
  * Updated Moonshot’s model catalog with new Kimi K2 variants and refreshed capabilities/descriptions.
* **Bug Fixes**
  * Improved Moonshot handling for unknown models to prevent incorrect token limits and pricing.
  * Refined Moonshot streaming, usage reporting, tool-call event behavior, and cache integration for more accurate updates.
* **Tests**
  * Expanded Moonshot coverage for model fetching, caching, router model aggregation, streaming, usage metrics, and tool-call streaming behavior.

### CODE REVIEW

**Comment by navedmerchant** [comment]
Thanks for this! Please bump coverage @grizmin

**Comment by grizmin** [comment]
> Thanks for this! Please bump coverage @grizmin

The remaining gaps are either UI components (Moonshot.tsx, useSelectedModel.ts) or minor edge cases. IMO all meaningful tests are in place.

File: webview-ui/src/components/settings/providers/Moonshot.tsx [hunk 1]
```diff
@@ -1,22 +1,62 @@
...
+					}
+				}
+			}
+		}
+
+		window.addEventListener("message", handleMessage)
+		return () => {
+			window.removeEventListener("message", handleMessage)
+		}
+	}, [refreshStatus, queryClient])
```
Comment (edelauna) [inline webview-ui/src/components/settings/providers/Moonshot.tsx:59]: This message listener drives a fairly involved state machine (idle/loading/success/error) with a same-tick race guard between the error and success messages. I don't see a test file for this component anywhere in the repo — is there one planned, particularly for the race case where an error arrives just before a stale `routerModels` success message?

File: src/api/providers/fetchers/modelCache.ts [hunk 2]
```diff
@@ -217,6 +220,9 @@ async function fetchModelsFromProvider(options: GetModelsOptions): Promise<Model
 		case "deepseek":
 			models = await getDeepSeekModels(options.baseUrl, options.apiKey)
 			break
+		case "moonshot":
+			models = await getMoonshotModels(options.baseUrl, options.apiKey)
```
Comment (edelauna) [inline src/api/providers/fetchers/modelCache.ts:228]: The existing test suite table-tests the `openrouter`/`requesty`/`litellm` branches of this switch by name — would it be worth adding a `moonshot` case here too, given it's a new provider going through the same dispatch?

File: webview-ui/src/components/ui/hooks/useSelectedModel.ts [hunk 3]
```diff
@@ -252,9 +252,13 @@ function getSelectedModel({
...
-			const info = moonshotModels[id as keyof typeof moonshotModels]
-			return { id, info }
+			const availableModels = routerModels.moonshot
+				? { ...moonshotModels, ...routerModels.moonshot }
+				: moonshotModels
+			const id = getValidatedModelId(apiConfiguration.apiModelId, availableModels, defaultModelId)
+			const routerInfo = routerModels.moonshot?.[id]
+			const staticInfo = moonshotModels[id as keyof typeof moonshotModels]
+			return { id, info: routerInfo ?? staticInfo }
 		}
```
Comment (edelauna) [inline webview-ui/src/components/ui/hooks/useSelectedModel.ts:263]: This mirrors the `deepseek` case right above it, which is good for consistency — but I don't see either case (or `getValidatedModelId` generally) exercised in the test file for a moonshot-shaped input. Worth adding a case here alongside the deepseek one?

File: src/api/providers/moonshot.ts [hunk 4]
```diff
@@ -5,30 +5,47 @@ import type { ApiHandlerOptions } from "../../shared/api"
...
+		const defaultInfo = moonshotModels[moonshotDefaultModelId]
+		return {
+			...defaultInfo,
+			maxTokens: undefined,
+			inputPrice: undefined,
+			outputPrice: undefined,
+			cacheReadsPrice: undefined,
+			cacheWritesPrice: undefined,
+		}
 	}
```
Comment (edelauna) [inline src/api/providers/moonshot.ts:46]: This strips pricing for unknown model IDs, which makes sense, but it also zeroes `maxTokens` from the default model's info — is that intentional? The doc comment above only mentions pricing being stripped.

File: src/api/providers/moonshot.ts [hunk 5]
```diff
@@ -43,34 +60,27 @@ export class MoonshotHandler extends OpenAICompatibleHandler {
...
 	 * Moonshot requires max_tokens parameter to be sent.
 	 */
-	protected override getMaxOutputTokens(): number | undefined {
-		const modelInfo = this.config.modelInfo
-		// Moonshot always requires max_tokens
-		return this.options.modelMaxTokens || modelInfo.maxTokens || undefined
+	protected override addMaxTokensIfNeeded(
+		requestOptions:
+			| import("openai/resources/chat/completions").ChatCompletionCreateParamsStreaming
+			| import("openai/resources/chat/completions").ChatCompletionCreateParamsNonStreaming,
```
Comment (edelauna) [inline src/api/providers/moonshot.ts]: Elsewhere in this codebase (e.g. `openai.ts`, `deepseek.ts`) the `ChatCompletionCreateParams*` types come from a top-level `import OpenAI from "openai"` + `OpenAI.Chat.Completions.X`, rather than an inline `import(...)` type. Any reason to diverge here, or just an oversight?

File: src/api/providers/fetchers/moonshot.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+					description: `Moonshot model: ${modelId}`,
+				}
+			}
+		}
+
+		return models
+	} finally {
+		clearTimeout(timeoutId)
+	}
+}
```
Comment (edelauna) [inline src/api/providers/fetchers/moonshot.ts:89]: This is nearly identical to `getDeepSeekModels` (timeout/AbortController setup, error handling, null-prototype merge loop) — just the default URL, unknown-model fallback, and log prefix differ. Worth factoring into a shared helper now that there are two, or defer until a third provider needs the same shape?

**Review by edelauna — CHANGES_REQUESTED** [review]
Thanks for this update! Could you also add UI tests (there should be prior art to pull from), since the ui relies on custom state machine.

Also before merge, can we update the documentation as well, see this issue: https://github.com/Zoo-Code-Org/Zoo-Code-Docs/issues/12

File: src/api/providers/moonshot.ts (hunk 5 above)
Comment (grizmin) [reply src/api/providers/moonshot.ts]: Replaced with OpenAI.Chat.Completions.X

File: webview-ui/src/components/settings/providers/Moonshot.tsx (hunk 1 above)
Comment (grizmin) [reply webview-ui/src/components/settings/providers/Moonshot.tsx:59]: Created Moonshot.spec.tsx with 10 tests including race condition guard

File: src/api/providers/fetchers/modelCache.ts (hunk 2 above)
Comment (grizmin) [reply src/api/providers/fetchers/modelCache.ts:228]: Added getMoonshotModels mock and dispatch test

File: src/api/providers/moonshot.ts (hunk 4 above)
Comment (grizmin) [reply src/api/providers/moonshot.ts:46]: Yes, intentional. Prevents sending guessed max_tokens for unknown models. The doc comment mentions pricing but maxTokens is stripped for the same reason - IMO unknown models shouldnt get guessed API parameters.

File: webview-ui/src/components/ui/hooks/useSelectedModel.ts (hunk 3 above)
Comment (grizmin) [reply webview-ui/src/components/ui/hooks/useSelectedModel.ts:263]: Added 4 tests: default model, router model override, unknown model fallback, and getValidatedModelId behavior.

File: src/api/providers/fetchers/moonshot.ts (hunk 6 above)
Comment (grizmin) [reply src/api/providers/fetchers/moonshot.ts:89]: Deferring the shared helper refactor until a third provider needs the same shape. Besides, it's out of scope for this feature.

**Review by edelauna — APPROVED** [review]
Thanks for the contribution and addressing the feedback.
